src/main/java/edu/cornell/kfs/module/purap/batch/service/impl/JaggaerUploadFileServiceImpl.java

```diff
@@ -1,42 +1,220 @@
+
+        boolean successfulCall = false;
+        int numberOfAttempts = 1;
+
+        while (!successfulCall && numberOfAttempts < getMaximumNumberOfRetries()) {
+            Client client = getClient();
+            WebTarget target = client.target(buildSupplierUploadURI());
+            Invocation.Builder requestBuilder = target.request();
+
+            disableRequestChunkingIfNecessary(client, requestBuilder);
+            
+            try (Response response = requestBuilder.accept(MediaType.APPLICATION_XML)
+                    .post(Entity.text(buildPostingStringFromJaggaerFile(jaggaerXmlFileName)));) {
+                String responseString = response.readEntity(String.class);
+
+                if (response.getStatusInfo().getFamily() == Response.Status.Family.SUCCESSFUL) {
+                    processSuccessfulResponse(results, responseString);
+                    successfulCall = true;
+                } else {
+                    processUnsuccessfulResponse(results, numberOfAttempts, responseString, response.getStatus());
+                    numberOfAttempts++;
+                }
+            }
+        }
+        return results;
+    }
+
+    private int getMaximumNumberOfRetries() {
+        String countString = getParameterValueString(CUPurapParameterConstants.JAGGAER_UPLOAD_RETRY_COUNT);
+        return Integer.valueOf(countString);
+    }
+
+    private String buildPostingStringFromJaggaerFile(String jaggaerXmlFileName) {
+        File jaggaerFile = new File(jaggaerXmlFileName);
+        String contentToPost;
+        try {
+            contentToPost = FileUtils.readFileToString(jaggaerFile, Charset.defaultCharset());
+        } catch (IOException e) {
+            LOG.error("buildPostingStringFromJaggaerFile, unable to read file " + jaggaerXmlFileName, e);
+            throw new RuntimeException(e);
         }
-        fileStorageService.removeDoneFiles(fileNamesToUpload);
+        LOG.debug("buildPostingStringFromJaggaerFile, about to post {}", contentToPost);
+        return contentToPost;
     }
 
-    /*
-     * @todo implement this with KFSPTS-28268
-     */
-    protected void uploadFile(byte[] uploadFileData) {
-        LOG.info("uploadFile, will eventually send a file with {} bytes of data to Jaggaer",
-                ArrayUtils.getLength(uploadFileData));
+    private void processSuccessfulResponse(JaggaerUploadFileResultsDTO results, String responseString) {
+        LOG.info("processSuccessfulResponse, responseString {}", responseString);
+        SupplierResponseMessage responseMessage = buildSupplierResponseMessage(responseString);
+
+        results.setResponseCode(responseMessage.getStatus().getStatusCode());
+        results.setMessage(responseMessage.getStatus().getStatusText());
+        results.setFileProcessedByJaggaer(true);
+
+        if (responseMessage.getStatus().getErrorMessages().isEmpty()) {
+            results.setErrorMessage(StringUtils.EMPTY);
```

Comment:
Yes, that is intentional.  The goal was to have the last try be the results that are reported.  Either a success will happen or it will fail.  The report will note however it failed on the last try.  The job log should have details about each failure if a developer needs to review what happened.